autoload/neoformat.vim

```diff
@@ -31,117 +31,123 @@ function! s:neoformat(bang, user_input, start_line, end_line) abort
         let &filetype = len(inputs) > 1 ? inputs[0] : a:user_input
     endif
 
-    let filetype = s:split_filetypes(&filetype)
-
+    let filetypes = s:split_filetypes(&filetype)
     let using_user_passed_formatter = (!empty(a:user_input) && !a:bang)
                 \ || (len(inputs) > 1 && a:bang)
 
     if using_user_passed_formatter
-        let formatters = len(inputs) > 1 ? [inputs[1]] : [a:user_input]
+        let filetype = s:split_filetype(&filetype)
+        let formatters = len(inputs) > 1 ? [[filetype, [inputs[1]]]] : [[filetype, [a:user_input]]]
     else
-        let formatters = s:get_enabled_formatters(filetype)
+        let formatters = []
+        for filetype in filetypes
+            call add(formatters, [filetype, s:get_enabled_formatters(filetype)])
+        endfor
         if formatters == []
-            call neoformat#utils#msg('formatter not defined for ' . filetype . ' filetype')
+            call neoformat#utils#msg('formatter not defined for ' . &filetype . ' filetype')
             return s:basic_format()
         endif
     endif
 
     let formatters_failed = []
     let formatters_changed = []
-    for formatter in formatters
-
-        if &formatprg != '' && split(&formatprg)[0] ==# formatter
-                    \ && neoformat#utils#var('neoformat_try_formatprg')
-            call neoformat#utils#log('using formatprg')
-            let fmt_prg_def = split(&formatprg)
-            let definition = {
-                    \ 'exe': fmt_prg_def[0],
-                    \ 'args': fmt_prg_def[1:],
-                    \ 'stdin': 1,
-                    \ }
-        elseif exists('b:neoformat_' . filetype . '_' . formatter)
-            let definition = b:neoformat_{filetype}_{formatter}
-        elseif exists('g:neoformat_' . filetype . '_' . formatter)
-            let definition = g:neoformat_{filetype}_{formatter}
-        elseif s:autoload_func_exists('neoformat#formatters#' . filetype . '#' . formatter)
-            let definition =  neoformat#formatters#{filetype}#{formatter}()
-        else
-            call neoformat#utils#log('definition not found for formatter: ' . formatter)
-            if using_user_passed_formatter
-                call neoformat#utils#msg('formatter definition for ' . a:user_input . ' not found')
+    for filetype_formatters in formatters
+        let filetype = filetype_formatters[0]
```

Comment:
I'm wondering if we should enable multiple file formatters conditionally

My concern is that someone could have a filetype like `typescript.jsx` or something and the first one works but the second one fails

or maybe the two formatters are incompatible and there is fighting between them

I'm curious what typical multi-file filetypes are common in the ecosystem

Reply:
It's hard to tell. Generally I would expect that if filetype is `typescript.jsx` then formatters for all filetypes should work correctly. In the same vein, it's uncertain whether all typescript formatters can handle jsx within file (or any other subfiletype that I might have invented), and yet currently NeoFormat will treat it as valid typescript file, which it might be not. I think it's generally a good assumption.

I think NeoFormat already has mechanisms to fulfil most common scenarios. User can specify per filetype whether one or more formatters should be used, and it is also possible to disable formatter. If I added option to disable search for subfiletype formatters with global or buffer level variable, combination of these three variables should be flexible enough to enable or disable this feature for specific filetypes. It could be disabled by default to maintain existing behaviour.
It's worth considering if NeoFormat should make additional checks like `g:neoformat_enabled_typescript_jsx = ['prettier']`. I think that is not necessary, it's very much possible to achieve that by setting `b:neoformat_enabled_filetype` in filetype plugin.

I wonder whether formatters should be de-duplicated. Let's stick to `typescript.jsx`, and suppose both filetypes have `prettier` definition. Should I run only first formatter, or run prettier twice? Perhaps prettier definition for `jsx` is different, e.g. different flags are passed to the command, or user has different settings for `jsx` prettier formatter, let say they wish different indentation width. This scenario could also be handled with proper setting enabled formatters in filetype plugin, though.

In conclusion, I think a variable to enable this behaviour should be sufficient. Disabling it by default will prevent users from stumbling upon weird combinations that wouldn't work without some manual tweaking.

Reply:
@sbdchd So, what do you think about that?

Reply:
OK, so I added option to enable using formatters for all subfiletypes. By default, Neoformat works as before, only first subfiletype is being considered.